### PR TITLE
qemu.test: check_memory_leak on windows guest after run command "guest-info" many times.

### PR DESCRIPTION
**Pull request by XueqiangWei**
Signed-off-by: Xueqiang Wei xuwei@redhat.com

### CODE REVIEW

**Comment by XueqiangWei** [comment]
ID 1377084

**Comment by XueqiangWei** [comment]
base on #706

**Comment by XueqiangWei** [comment]
smp_8.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_blk.up.virtio_net.Win8.i386.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.check_memory_leak: PASS

File: qemu/tests/qemu_guest_agent.py [hunk 1]
```diff
@@ -460,6 +460,54 @@ def gagent_check_set_time(self, test, params, env):
...
+        """
+        repeat execute "guest-info" command to guest agent, check memory
+        usage of the qemu-ga
+
+        :param test: kvm test object
+        :param params: Dictionary with the test parameters
+        :param env: Dictionary with test environment.
+        """
+
+        timeout = float(params.get("login_timeout", 240))
```
Comment (xutian) [inline qemu/tests/qemu_guest_agent.py]: how about define a new function like,

def _get_mem_used(session, cmd):
     out = session.cmd_output(cmd)
     try:
          used_mem = int(out.split(" ")[-2].replace(",", ""))
          return used_mem
     except Exception:
          raise exceptions.TestError("Get invalid memory usage by cmd '%s' (%s)" % (cmd, out))

Comment (XueqiangWei) [inline qemu/tests/qemu_guest_agent.py]: thx, will update

**Comment by XueqiangWei** [comment]
smp_8.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_blk.up.virtio_net.Win8.i386.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.check_memory_leak: PASS

File: qemu/tests/qemu_guest_agent.py [hunk 2]
```diff
@@ -460,6 +461,65 @@ def gagent_check_set_time(self, test, params, env):
...
+                      test_command, logging.info)
+        memory_usage = self._get_mem_used(session, memory_usage_cmd)
+        session.close()
+        repeats = int(params.get("repeats", 1))
+        for i in range(repeats):
+            error.context("execute '%s' %s times" % (test_command, i+1),
+                          logging.info)
+            return_msg = self.gagent.guest_info()
+            logging.info(str(return_msg))
+        self.vm.verify_alive()
```
Comment (suqinhuang) [inline qemu/tests/qemu_guest_agent.py]: hi xuwei,

why need to verify_alive and login again here

Comment (XueqiangWei) [inline qemu/tests/qemu_guest_agent.py]: it repeats run "guest-info" 1000000 times, need to verify the guest is alive after run cmd. And then  get memory usage again, compare with the raw memory usage that before run cmd.

File: qemu/tests/qemu_guest_agent.py [hunk 3]
```diff
@@ -460,6 +461,65 @@ def gagent_check_set_time(self, test, params, env):
...
+        error.context("get the memory usage of qemu-ga after run '%s'" %
+                      test_command, logging.info)
+        session = self.vm.wait_for_login(timeout=timeout)
+        memory_usage_ = self._get_mem_used(session, memory_usage_cmd)
+        session.close()
+        # less than 20K is acceptable.
+        if memory_usage_ - memory_usage > 20:
```
Comment (CongLi) [inline qemu/tests/qemu_guest_agent.py]: Could you use name like "mem_usage_before" and "mem_usage_after"?
memory_usage_ and memory_usage are really similar.

File: qemu/tests/qemu_guest_agent.py [hunk 4]
```diff
@@ -460,6 +461,65 @@ def gagent_check_set_time(self, test, params, env):
...
+            raise error.TestError("The memory usages are different, "
```
Comment (CongLi) [inline qemu/tests/qemu_guest_agent.py]: Should be TestFail I think.

Comment (XueqiangWei) [inline qemu/tests/qemu_guest_agent.py]: I test many times. It is OK. memory_usage_after is greater than memory_usage_before winthin 20kb is ok.

Comment (CongLi) [inline qemu/tests/qemu_guest_agent.py]: Misunderstanding?
I mean shoud use error.TestFail() here, TestError() is usually for env prepare failed.

File: qemu/tests/qemu_guest_agent.py [hunk 5]
```diff
@@ -460,6 +461,65 @@ def gagent_check_set_time(self, test, params, env):
...
+            logging.info(str(return_msg))
+        self.vm.verify_alive()
+        error.context("get the memory usage of qemu-ga after run '%s'" %
+                      test_command, logging.info)
+        session = self.vm.wait_for_login(timeout=timeout)
+        memory_usage_after = self._get_mem_used(session, memory_usage_cmd)
+        session.close()
+        # less than 20K is acceptable.
+        if memory_usage_after - memory_usage_before > 20:
+            raise error.TestError("The memory usages are different, "
```
Comment (xutian) [inline qemu/tests/qemu_guest_agent.py]: agree with @CongLi , It should be TestFail here

File: qemu/tests/qemu_guest_agent.py (hunk 4 above)
Comment (XueqiangWei) [inline qemu/tests/qemu_guest_agent.py]: got it, thx

File: qemu/tests/qemu_guest_agent.py (hunk 5 above)
Comment (XueqiangWei) [inline qemu/tests/qemu_guest_agent.py]: ok, thx, will update

File: qemu/tests/cfg/qemu_guest_agent.cfg [hunk 6]
```diff
@@ -54,6 +54,12 @@
         - check_get_time:
             gagent_check_type = get_time
             get_guest_time_cmd = python -c "import time; print int(time.time())"
+        - check_memory_leak:
+            only Windows
+            gagent_check_type = memory_leak
+            repeats = 1000000
```
Comment (xutian) [inline qemu/tests/cfg/qemu_guest_agent.cfg]: repeat 1000000 times ! are sure this case can be finished be test_timeout ?

Comment (XueqiangWei) [inline qemu/tests/cfg/qemu_guest_agent.cfg]: Yes, I tested it twice. It takes about 3 hours.

File: qemu/tests/qemu_guest_agent.py [hunk 7]
```diff
@@ -460,6 +461,65 @@ def gagent_check_set_time(self, test, params, env):
...
+            return_msg = self.gagent.guest_info()
+            logging.info(str(return_msg))
+        self.vm.verify_alive()
+        error.context("get the memory usage of qemu-ga after run '%s'" %
+                      test_command, logging.info)
+        session = self.vm.wait_for_login(timeout=timeout)
+        memory_usage_after = self._get_mem_used(session, memory_usage_cmd)
+        session.close()
+        # less than 20K is acceptable.
+        if memory_usage_after - memory_usage_before > 20:
```
Comment (ngu-niny) [inline qemu/tests/qemu_guest_agent.py]: 20K is not enough, once test failed:

22:22:03 ERROR| TestFail: The memory usages are different, before run command is 7624kb and after run command is 7744kb    [context: get the memory usage of qemu-ga after run 'guest-info']

Comment (XueqiangWei) [inline qemu/tests/qemu_guest_agent.py]: thanks ngu-niny, will update to 500kb.

**Comment by weliao** [comment]
2016-10-20 14:17:22,069 test             L0595 INFO | PASS 1-smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_blk.up.virtio_net.Win8.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.check_memory_leak

Acked-by: Wei Liao weliao@redhat.com

**Comment by CongLi** [comment]
Thanks @XueqiangWei @weliao @xutian @suqinhuang , merge.

**Pull request merged**